base/services/w32time/w32time.c

```diff
@@ -0,0 +1,418 @@
+            {
+                /*
+                 * We successfully enabled it, we're permitted to change the time.
+                 */
+                Ret = SetSystemTime(lpSystemTime);
+
+                /*
+                 * For the sake of security, restore the previous status again
+                 */
+                if (previouspriv.PrivilegeCount > 0)
+                {
+                    AdjustTokenPrivileges(hToken,
+                                          FALSE,
+                                          &previouspriv,
+                                          0,
+                                          NULL,
+                                          0);
+                }
+            }
+        }
+        CloseHandle(hToken);
+    }
+
+    return Ret;
+}
+
+
+/*
+ * NTP servers state the number of seconds passed since
+ * 1st Jan, 1900. The time returned from the server
+ * needs adding to that date to get the current Gregorian time
+ */
+static DWORD
+UpdateSystemTime(ULONG ulTime)
+{
+    FILETIME ftNew;
+    LARGE_INTEGER li;
+    SYSTEMTIME stNew;
+
+    /* Time at 1st Jan 1900 */
+    stNew.wYear = 1900;
+    stNew.wMonth = 1;
+    stNew.wDay = 1;
+    stNew.wHour = 0;
+    stNew.wMinute = 0;
+    stNew.wSecond = 0;
+    stNew.wMilliseconds = 0;
+
+    /* Convert to a file time */
+    if (!SystemTimeToFileTime(&stNew, &ftNew))
+    {
+        return GetLastError();
+    }
+
+    /* Add on the time passed since 1st Jan 1900 */
+    li = *(LARGE_INTEGER *)&ftNew;
+    li.QuadPart += (LONGLONG)10000000 * ulTime;
+    ftNew = * (FILETIME *)&li;
+
+    /* Convert back to a system time */
+    if (!FileTimeToSystemTime(&ftNew, &stNew))
+    {
+        return GetLastError();
+    }
+
+    if (!SystemSetTime(&stNew))
+    {
+        return GetLastError();
+    }
+
+    return ERROR_SUCCESS;
+}
+
+
+static DWORD
+GetIntervalSetting(VOID)
+{
+    HKEY hKey;
+    DWORD dwData;
+    DWORD dwSize = sizeof(dwData);
+    LONG lRet;
+
+    dwData = 0;
+    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                      L"SYSTEM\\CurrentControlSet\\Services\\W32Time\\TimeProviders\\NtpClient",
+                      0,
+                      KEY_QUERY_VALUE,
+                      &hKey) == ERROR_SUCCESS)
+    {
+        /* This key holds the update interval in seconds
+         * It is useful for testing to set it to a value of 10 (Decimal)
+         * This will cause the clock to try and update every 10 seconds
+         * So you can change the time and expect it to be set back correctly in 10-20 seconds
+         */
+        lRet = RegQueryValueExW(hKey,
+                                L"SpecialPollInterval",
+                                NULL,
+                                NULL,
+                                (LPBYTE)&dwData,
+                                &dwSize);
+        RegCloseKey(hKey);
+    }
+
+    if (lRet != ERROR_SUCCESS)
+        return 0;
+    else
+        return dwData;
+}
+
+
+int
+SetTime(VOID)
```

Comment:
Should we make this return BOOL/TRUE/FALSE, or are there some added semantics that might come later where `int` would be useful?
We could also just have it directly return `HRESULT`, in which case we'd gain the ability to return more specific error codes from W32TimeSyncNow.

Reply:
My intent was to leave this to allow reporting of multiple errors in the future. I have not been able to determine what these might be yet though even after doing some Internet research. Next I was going to try and break this and see what kinds of errors Windows reports and try and make our function report these errors.